File: 50dias/python/47.py
import os

prodStringSize = 20
codigos = [100, 101, 102, 103, 104, 105]
produtos = {codigos[0]: {'nome': 'Cachorro Quente',
                         'preco': 1.20},
            codigos[1]: {'nome': 'Bauru Simples',
                         'preco': 1.30},
            codigos[2]: {'nome': 'Bauru com ovo',
                         'preco': 1.50},
            codigos[3]: {'nome': 'Hamburguer',
                         'preco': 1.20},
            codigos[4]: {'nome': 'Cheeseburguer',
                         'preco': 1.30},
            codigos[5]: {'nome': 'Refrigerante',
                         'preco': 1.00}}

pedidosTotais = {}
codigoPedido = 0

def ImprimeCardapio():
    print('------------------------------------------')
    print('  Produto             |  Codigo  |  Preco ')
    print('------------------------------------------')
    for codigo in codigos:
        produto = produtos[codigo]
        nome = produto['nome'].ljust(prodStringSize)
        preco = produto['preco']
        msg = '  {nome}|  {cod}     |  R$ {preco:.2f}'.format(nome=nome,
                                                              cod=codigo,
                                                              preco=preco)
        print(msg)
    print('------------------------------------------')

def ImprimePedido(pedidoAtual):
    precoTotal = 0.00
    print('==========================================')
    print('               Seu pedido                 ')
    print('==========================================')
    print('  Produto             |  Quant   |  Preco ')
    print('------------------------------------------')
    for codigo in codigos:
        quantidade = pedidoAtual[codigo]
        if quantidade > 0:
            produto = produtos[codigo]
            nome = produto['nome'].ljust(prodStringSize)
            preco = produto['preco']
            msg = '  {nome}|  {quant}     |  R$ {preco:.2f}'.format(nome=nome,
                                                                    quant=quantidade,
                                                                    preco=quantidade * preco)
            print(msg)
            precoTotal += quantidade * preco

    print('------------------------------------------')
    print('  Preco Total: R$ {preco:.2f}'.format(preco=precoTotal))
    print('------------------------------------------')

def ImprimeMenuPedidos(pedidoAtual):
    ImprimeCardapio()
    ImprimePedido(pedidoAtual)

def RegistraPedido(pedidoAtual, codigo, quantidade):
    pedidoAtual[codigo] += quantidade

def LeOpcoesDoUsuario():
    opcao = int(input('Digite o codigo do produto (0 para sair): '))
    quantidade = -1
    if opcao != 0:
        quantidade = int(input('Digite a quantidade: '))
    return opcao, quantidade

def ValidaInformacoes(opcao, quantidade):
    produtoExiste = opcao in codigos
    quantidadeValida = quantidade > 0
    if not produtoExiste:
        print('\n  Codigo de produto nao encontrado!\n')
    elif not quantidadeValida:
        print('\n  Quantidade digitada eh invalida!\n')

    return produtoExiste and quantidadeValida

def CriaPedido():
    pedido = {codigos[0]: 0,
              codigos[1]: 0,
              codigos[2]: 0,
              codigos[3]: 0,
              codigos[4]: 0,
              codigos[5]: 0}
    return pedido

def LePedidoDoUsuario():
    global codigoPedido
    codigoPedido += 1
    nomeUsuario = input('Digite o seu nome: ')
    pedidoAtual = CriaPedido()
    opcao = -1
    os.system('cls')
    while opcao != 0:
        ImprimeMenuPedidos(pedidoAtual)
        opcao, quantidade = LeOpcoesDoUsuario()
        if opcao != 0:
            os.system('cls')
            if ValidaInformacoes(opcao, quantidade):
                RegistraPedido(pedidoAtual, opcao, quantidade)

    pedidosTotais[codigoPedido] = {'nome': nomeUsuario,
                                   'pedido': pedidoAtual}
                                   
    print('\n  O seu pedido é o número: {}'.format(codigoPedido))
    print('  Efetue o pagamento no caixa.')
    print('  Muito obrigado pelo seu pedido!')
    _ = input('\nTecle ENTER para sair!')

def ImprimeMenuSistema():
    print('==========================================')
    print('        Lancheria do Juquinha             ')
    print('==========================================')
    print('  1 - Fazer pedido ')
    print('  9 - Sair ')
    print('------------------------------------------')
    opcao = -1
    while opcao != 1 and opcao != 9:
        opcao = int(input('Digite a opcao: '))
    return opcao

def ImprimePedidosDoDia():
    print(pedidosTotais)
    for codigoPedido, pedidoUsuario in pedidosTotais.items():
        print(' Pedido n: {}'.format(codigoPedido))
        print(' Cliente: {}'.format(pedidoUsuario['nome']))
        ImprimePedido(pedidoUsuario['pedido'])

def main():
    opcaoSistema = -1
    while opcaoSistema != 9:
        opcaoSistema = ImprimeMenuSistema()
        if opcaoSistema != 9:
            LePedidoDoUsuario()

    ImprimePedidosDoDia()

main()

File: 50dias/cpp/47.cpp
#include <iostream>
#include <vector>
#include <map>
#include <string>
#include <iomanip>
#include <list>
#include <ctime>
#include <conio.h>

using namespace std;

const int c_tamanhoProduto = 20;
const vector<int> codigos = { 100, 101, 102, 103, 104, 105 };
int codigoGeralPedido = 0;

map<int, float> precos = { {codigos[0], 1.20},
						   {codigos[1], 1.30},
						   {codigos[2], 1.50},
						   {codigos[3], 1.20},
						   {codigos[4], 1.30},
						   {codigos[5], 1.00} };

map<int, string> produtos = { {codigos[0], "Cachorro Quente"},
							  {codigos[1], "Bauru Simples"},
							  {codigos[2], "Bauru com ovo"},
							  {codigos[3], "Hamburguer"},
							  {codigos[4], "Cheesburguer"},
							  {codigos[5], "Refrigerante"} };


class PedidoUsuario {
public:
	PedidoUsuario(string NomeUsuario, int codigoPedido) :
		m_nomeUsuario(NomeUsuario),
		m_codigoPedido(codigoPedido)
	{
		m_data = time(0);
	}

	void RegistraPedido(int codigo, int quantidade) {
		m_pedido[codigo] += quantidade;
	}

	void ImprimePedido() const {
		cout << "==============================================" << endl;
		cout << "                Seu pedido                    " << endl;
		cout << "==============================================" << endl;
		cout << "  Produto             |  Quant  |  Preco" << endl;
		cout << "----------------------------------------------" << endl;
		float precoTotal = 0;
		for (auto item : m_pedido) {
			int codigo = item.first;
			int quantidade = item.second;
			if (quantidade > 0) {
				string nome = produtos[codigo];
				float preco = precos[codigo];
				cout << "  " << nome << setw(c_tamanhoProduto - nome.size()) << " ";
				cout << "|  " << setw(3) << quantidade << "    |  R$ ";
				cout << std::fixed << std::setprecision(2);
				cout << quantidade * preco << endl;
				precoTotal += quantidade * preco;
			}
		}
		cout << "----------------------------------------------" << endl;
		cout << "  Preco Total: R$ " << precoTotal << endl;
		cout << "----------------------------------------------" << endl;
	}

	void ImprimeRelatorioPedido() const {
		char* dataHora = ctime(&m_data);

		cout << endl;
		cout << " Pedido n: " << m_codigoPedido << endl;
		cout << " Cliente: " << m_nomeUsuario << endl;
		cout << " Data: " << dataHora << endl;
		ImprimePedido();
	}

private:
	int m_codigoPedido;
	string m_nomeUsuario;
	time_t m_data;
	map<int, int> m_pedido = { {codigos[0], 0},
								 {codigos[1], 0},
								 {codigos[2], 0},
								 {codigos[3], 0},
								 {codigos[4], 0},
								 {codigos[5], 0} };

};

list<PedidoUsuario> pedidosTotais;

void ImprimeCardapio() {
	cout << "----------------------------------------------" << endl;
	cout << "  Produto             |  Codigo  |  Preco" << endl;
	cout << "----------------------------------------------" << endl;
	for (auto codigo : codigos) {
		string nome = produtos[codigo];
		float preco = precos[codigo];
		cout << "  " << nome << setw(c_tamanhoProduto - nome.size()) << " ";
		cout << "|  " << codigo << "     |  R$ ";
		cout << std::fixed << std::setprecision(2);
		cout << preco << endl;
	}
	cout << "----------------------------------------------" << endl;
}

void LeOpcoesDoUsuario(int& opcao, int& quantidade) {
	cout << "Digite o codigo do produto (0 para sair): ";
	cin >> opcao;
	quantidade = -1;
	if (opcao != 0) {
		cout << "Digite a quantidade: ";
		cin >> quantidade;
	}
}

bool ValidaInformacoes(int opcao, int quantidade) {
	bool produtoExiste = produtos.find(opcao) != produtos.end();
	bool quantidadeValida = quantidade > 0;
	if (!produtoExiste) {
		cout << endl << "  Codigo de produto nao encontrado!";
		cout << endl << endl;
	}
	else {
		if (!quantidadeValida) {
			cout << endl << "  Quantidade digitada eh invalida!";
			cout << endl << endl;
		}
	}

	return produtoExiste && quantidadeValida;
}

void LePedidoDoUsuario() {
	codigoGeralPedido++;
	string nomeUsuario;
	cout << "Digite o seu nome: ";
	getline(cin, nomeUsuario);

	PedidoUsuario pedidoAtual(nomeUsuario, codigoGeralPedido);

	int opcao, quantidade;
	system("cls");
	do {
		ImprimeCardapio();
		pedidoAtual.ImprimePedido();

		LeOpcoesDoUsuario(opcao, quantidade);

		if (opcao != 0) {
			system("cls");
			if (ValidaInformacoes(opcao, quantidade)) {
				pedidoAtual.RegistraPedido(opcao, quantidade);
			}
		}

	} while (opcao != 0);

	pedidosTotais.push_back(pedidoAtual);
	cout << "\n  O seu pedido e o numero:" << codigoGeralPedido << endl;
	cout << "  Efetue o pagamento no caixa." << endl;
	cout << "  Muito obrigado pelo seu pedido!" << endl;
	cout << "\nTecle ENTER para sair!";
	getch();
}

int ImprimeMenuSistema() {
	cout << "==========================================" << endl;
	cout << "        Lancheria do Juquinha             " << endl;
	cout << "==========================================" << endl;
	cout << "  1 - Fazer pedido " << endl;
	cout << "  9 - Sair " << endl;
	cout << "------------------------------------------" << endl;
	int opcaoSistema = 0;
	do {
		cout << "Digite a opcao: ";
		cin >> opcaoSistema;
		cin.ignore();
	} while (opcaoSistema != 1 and opcaoSistema != 9);
	return opcaoSistema;
}

void ImprimePedidosDoDia() {
	for (auto pedido : pedidosTotais) {
		pedido.ImprimeRelatorioPedido();
	}
}

int main()
{
	bool continuarSistema = false;
	int quantidade = 0;
	do {
		system("cls");
		int opcaoSistema = ImprimeMenuSistema();
		continuarSistema = opcaoSistema != 9;
		if (continuarSistema) {
			LePedidoDoUsuario();
		}
	} while (continuarSistema);

	ImprimePedidosDoDia();
	return 0;
}